src/esp/geo/VoxelGrid.h

```diff
@@ -0,0 +1,515 @@
+   */
+  template <typename T>
+  void addGrid(const std::string& gridName) {
+    std::string type;
+    // set type depending on the type of T
+    if (std::string(typeid(T).name()) == "i") {
+      type = "int";
+    } else if (std::string(typeid(T).name()) == "f") {
+      type = "float";
+    } else if (std::string(typeid(T).name()) == "b") {
+      type = "bool";
+    } else if (std::string(typeid(T).name()) == "N6Magnum4Math7Vector3IfEE") {
+      type = "vector3";
+    }
+
+    if (grids_.find(gridName) != grids_.end()) {
+      // grid exists, simply overwrite
+      Mn::Debug() << gridName << "exists, overwriting.";
+      Corrade::Containers::Array<char> cr_grid{Corrade::Containers::ValueInit,
+                                               gridSize() * sizeof(T)};
+      grids_[gridName].second = std::move(cr_grid);
+      grids_[gridName].first = type;
+      return;
+    }
+    Corrade::Containers::Array<char> cr_grid{Corrade::Containers::ValueInit,
+                                             gridSize() * sizeof(T)};
+    grids_.insert(
+        std::make_pair(gridName, std::make_pair(type, std::move(cr_grid))));
+  }
+
+  /**
+   * @brief Removes a grid and frees up memory.
+   * @param name The name of the grid to be removed.
+   */
+  void removeGrid(const std::string& gridName) {
+    assert(grids_.find(gridName) != grids_.end());
+    grids_.erase(gridName);
+  }
+
+  template <typename T>
+  Cr::Containers::StridedArrayView<3, T> getGrid(const std::string& gridName) {
+    std::string type = grids_[gridName].first;
+    unsigned long dims[3]{static_cast<unsigned long>(m_voxelGridDimensions[0]),
+                          static_cast<unsigned long>(m_voxelGridDimensions[1]),
+                          static_cast<unsigned long>(m_voxelGridDimensions[2])};
+    return Cr::Containers::StridedArrayView<3, T>{
+        Cr::Containers::arrayCast<T>(grids_[gridName].second), dims};
+  }
+
+  /**
+   * @brief Linearizes 3D voxel coordinates to a single value in order to
+   * directly access a voxel grid.
+   * @param coords The 3D voxel index to be linearized.
+   * @return The has value of the voxel index.
+   */
+  int hashVoxelIndex(const Mn::Vector3i& coords);
+
+  /**
+   * @brief Converts a hash value into a 3 dimensional coordinate.
+   * @param hash The hash value.
+   * @return The voxel index derived from the given hash value.
+   */
+  Mn::Vector3i reverseHash(const int hash);
+
+  /**
+   * @brief Checks to see if a given 3D voxel index is valid and does not go out
+   * of bounds.
+   * @param coords The voxel index.
+   * @return True if the voxel index is valid, false otherwise.
+   */
+  bool isValidIndex(const Mn::Vector3i& coords) const {
+    return coords[0] >= 0 && coords[1] >= 0 && coords[2] >= 0 &&
+           coords[0] < m_voxelGridDimensions[0] &&
+           coords[1] < m_voxelGridDimensions[1] &&
+           coords[2] < m_voxelGridDimensions[2];
```

Comment:
Most pperations done on scalars can be done on the vectors as well :) In this case:

```cpp
return (coords >= Mn::Vector3i{0}).all() && (coords < m_voxelGridDimensions).all();
```

(I was trying to figure out if there was a way to not need this function but failed. So at least this.)

Reply:
Bool vectors! This is neat.